### ISSUE TITLE
AI evals deep — au-delà des benchmarks publics

### ISSUE THREAD

**Issue by khalilbenaz** [khalilbenaz/khalilbenaz.github.io#197]
---
lede: Construire des evals robustes nécessite plus que des comparaisons LLM-as-judge. Patterns, pitfalls, outils — guide pratique.
lede_en: Building robust evals takes more than LLM-as-judge comparisons. Patterns, pitfalls, tools — practical guide.
title_en: AI evals deep — beyond public benchmarks
---

## Pourquoi les evals deviennent critiques

Les apps LLM en production sans evals = vol en aveugle. Vous changez un prompt, vous ne savez pas si vous régressez. Vous switchez de modèle, vous ne savez pas si la qualité tient. Vous ajoutez du retrieval, vous espérez que ça aide.

Construire un **dataset d'évaluation propre** + **métriques mesurables** = la fondation de toute app IA sérieuse.

## Au-delà des benchmarks publics

MMLU, HumanEval, GSM8K sont utiles pour comparer des modèles en absolu. Mais ils ne disent **rien** sur votre cas d'usage spécifique :

- Trop génériques.
- Contaminés (souvent dans les training sets).
- N'évaluent pas votre prompt ni votre retrieval.

Un eval set **fait pour votre app** est non-négociable.

## Patterns d'évaluation avancés

**1. Pairwise comparison**

Au lieu de noter "5/10", vous présentez deux réponses à un LLM-judge et demandez laquelle est meilleure. Plus fiable que scoring absolu :

```ts
const winner = await judge(`
  Question: ${question}
  Response A: ${responseA}
  Response B: ${responseB}
  Which is more helpful and accurate? Answer A or B.
`);
```

**2. Rubric-based scoring**

Décomposer en critères et noter chacun :

- Accuracy (0-5)
- Helpfulness (0-5)
- Safety (0-5)
- Style (0-5)

Plus précis que un score global. Identifie où le modèle pêche.

**3. Adversarial generation**

Utilisez un autre LLM pour **générer des cas limites** que votre app pourrait rater :

- Questions ambiguës.
- Requêtes dans un format inattendu.
- Tentatives de jailbreak.
- Inputs très longs ou très courts.

Vous trouvez des bugs avant les utilisateurs.

**4. User feedback loop**

Intégrer le feedback réel (thumbs up/down, edits utilisateur) dans votre eval set. Au bout d'un mois, votre eval reflète le vrai usage.

## Pitfalls classiques

- **Dataset trop petit** : <30 examples = bruit dominant. Viser 50-200 minimum.
- **Dataset déconnecté du trafic** : si votre eval ne reflète pas le real usage, vous optimisez le mauvais signal.
- **LLM-as-judge avec le même modèle** : biais. Toujours utiliser un modèle différent pour juger.
- **Pas de versioning** : eval qui change sans control = scores incomparables dans le temps.
- **Métrique unique** : combinez plusieurs métriques. Tendance à overfit sur une seule.

## Outils

- **Promptfoo** : CLI Node, déclaratif (YAML), simple et puissant.
- **Inspect** (UK AISI) : framework Python complet, gratuit, open source.
- **LangSmith** : payant, intégré LangChain.
- **LangFuse** : tracing + eval, open source.
- **DIY** : 100 lignes de Python suffisent pour démarrer simple.

## Workflow type

1. **Collect 50-100 real examples** depuis vos logs.
2. **Annoter** : ce que devrait être la réponse "idéale".
3. **Définir 2-3 métriques** (accuracy, format, latence).
4. **Run baseline** : score votre system actuel.
5. **Itérer** : changer le prompt, le modèle, le retrieval. Mesurer.
6. **Promote** seulement si l'eval s'améliore.

## Au-delà du LLM-as-judge

Pour des tâches très spécifiques, dépasser le LLM-judge :

- **Validation programmatique** : pour du JSON, schéma. Pour de la code generation, exécution unit tests.
- **Human evaluation** : pour des decisions critiques (médical, légal), humain valide un échantillon.
- **A/B tests live** : la métrique ultime = conversion utilisateur.

## Le réflexe pro

Pas de mise en prod sans baseline d'eval. Pas de changement de prompt sans re-run l'eval. Pas de switch de modèle sans comparaison. **Data-driven** appliqué à l'IA.

## Why evals become critical

LLM apps in production without evals = flying blind. Change a prompt, you don't know if you regress. Switch model, you don't know if quality holds. Add retrieval, you hope it helps.

Building a **clean evaluation dataset** + **measurable metrics** = the foundation of any serious AI app.

## Beyond public benchmarks

MMLU, HumanEval, GSM8K are useful to compare models in absolute. But they say **nothing** about your specific use case:

- Too generic.
- Contaminated (often in training sets).
- Don't evaluate your prompt or retrieval.

An eval set **made for your app** is non-negotiable.

## Advanced evaluation patterns

**1. Pairwise comparison**

Instead of scoring "5/10", present two responses to an LLM-judge and ask which is better. More reliable than absolute scoring:

```ts
const winner = await judge(`
  Question: ${question}
  Response A: ${responseA}
  Response B: ${responseB}
  Which is more helpful and accurate? Answer A or B.
`);
```

**2. Rubric-based scoring**

Decompose into criteria and score each:

- Accuracy (0-5)
- Helpfulness (0-5)
- Safety (0-5)
- Style (0-5)

More precise than a global score. Identifies where the model fails.

**3. Adversarial generation**

Use another LLM to **generate edge cases** your app might fail:

- Ambiguous questions.
- Requests in unexpected format.
- Jailbreak attempts.
- Very long or very short inputs.

You find bugs before users.

**4. User feedback loop**

Integrate real feedback (thumbs up/down, user edits) into your eval set. After a month, your eval reflects real usage.

## Classic pitfalls

- **Too-small dataset**: <30 examples = noise dominant. Aim for 50-200 minimum.
- **Dataset disconnected from traffic**: if your eval doesn't reflect real usage, you optimize the wrong signal.
- **LLM-as-judge with the same model**: bias. Always use a different model to judge.
- **No versioning**: eval changing without control = scores incomparable over time.
- **Single metric**: combine multiple metrics. Tendency to overfit on one.

## Tools

- **Promptfoo**: Node CLI, declarative (YAML), simple and powerful.
- **Inspect** (UK AISI): complete Python framework, free, open source.
- **LangSmith**: paid, LangChain-integrated.
- **LangFuse**: tracing + eval, open source.
- **DIY**: 100 lines of Python suffice for simple start.

## Typical workflow

1. **Collect 50-100 real examples** from your logs.
2. **Annotate**: what the "ideal" response should be.
3. **Define 2-3 metrics** (accuracy, format, latency).
4. **Run baseline**: score your current system.
5. **Iterate**: change the prompt, model, retrieval. Measure.
6. **Promote** only if eval improves.

## Beyond LLM-as-judge

For very specific tasks, go beyond LLM-judge:

- **Programmatic validation**: for JSON, schema. For code generation, execute unit tests.
- **Human evaluation**: for critical decisions (medical, legal), humans validate a sample.
- **Live A/B tests**: the ultimate metric = user conversion.

## The pro reflex

No production deploy without an eval baseline. No prompt change without re-running the eval. No model switch without comparison. **Data-driven** applied to AI.
